src/index.js

```diff
@@ -58,45 +58,59 @@ module.exports = class FaviconsWebpackPlugin {
 
     // Hook into the webpack compilation
     // to start the favicon generation
-    tap(compiler, 'make', 'FaviconsWebpackPlugin', (compilation, callback) => {
-      let faviconCompilation;
-      switch (this.getCurrentCompilationMode(compiler) ) {
-        case 'light':
-          faviconCompilation = this.generateFaviconsLight(compiler, compilation);
-          break;
-        case 'webapp':
-        default:
-          faviconCompilation = this.generateFaviconsWebapp(compiler, compilation);
-      }
-      
+    compiler.hooks.make.tapPromise('FaviconsWebpackPlugin', async compilation => {
+      const faviconCompilation = this.generateFavicons(compiler, compilation);
+
       // Hook into the html-webpack-plugin processing and add the html
-      tapHtml(compilation, 'FaviconsWebpackPlugin', (htmlPluginData, htmlWebpackPluginCallback) => {
-        faviconCompilation.then((tags) => {
-          if (this.options.inject(htmlPluginData.plugin)) {
-            const idx = (htmlPluginData.html + '</head>').search(/<\/head>/i);
-            htmlPluginData.html = [htmlPluginData.html.slice(0, idx), ...tags, htmlPluginData.html.slice(idx)].join('');
-          }
-          htmlWebpackPluginCallback(null, htmlPluginData);
-        }).catch(htmlWebpackPluginCallback);
-      });
+      const htmlWebpackPlugin = compiler.options.plugins
+        .map(({ constructor }) => constructor)
+        .find(({ name }) => name === 'HtmlWebpackPlugin');
+
+      if (htmlWebpackPlugin) {
+        htmlWebpackPlugin.getHooks(compilation).alterAssetTags.tapAsync('FaviconsWebpackPlugin', (htmlPluginData, htmlWebpackPluginCallback) => {
+          faviconCompilation.then((tags) => {
 
-      // Save the promise and execute the callback immediately to not block 
-      // the webpack build see the `afterCompile` FaviconsWebpackPlugin hook 
+            if (this.options.inject(htmlPluginData.plugin)) {
+              htmlPluginData.plugin.options.inject = true;
```

Comment:
Should we really overrule this setting?
`inject:true` is the default of the html-webpack-plugin so if it is set to false it is an explicit user choice.

Maybe we could output a warning if it is set to `false`?

Reply:
What happens if `htmlPluginData.plugin.options.inject === false`? 
Will it silently ignore the tags we add? I wonder if it shouldn't be a warning on `html-webpack-plugin` side instead if someone hooks to `alterAssetTags` while `inject` is false..

Reply:
This needs to be thought through though, because of the feature that allows one to set `faviconsPlugin.inject = 'force'` to explicitly overrule `htmlPlugin.inject`. If we remove this line it will break that feature.

Reply:
The `assetTags` can be used inside the template even if `inject` is set to `false`.

You just have to add the following line from the [html-webpack-plugin example](
https://github.com/jantimon/html-webpack-plugin/blob/35a154186501fba3ecddb819b6f632556d37a58f/examples/custom-insertion-position/index.ejs#L4):  
 `<%= htmlWebpackPlugin.tags.headTags %>`

Reply:
Got it, in this case I'd remove the override in the same PR that the `force` feature is removed, does that make sense?